### DISCUSSION TITLE
Adding a Table to a livewire component, and using that component in an Infolist

### DISCUSSION THREAD

**Discussion by clrockwell** [filamentphp/filament#8390] [Help] [answered]
I have the following defined on the Resource: 

**Update: I have updated the code based on feedback and current status**

**Final Update**: After making the changes that resulted in the state of the code being what is shown below, the last error was because I needed to define the `$primaryKey` on my model.  What isn't clear to me is that I needed to bind the record (using `:course=$getRecord()`) - that is likely due to me conflating the Livewire and FilamentPHP functionality as I stumble my way through understanding these two AWESOME tools.

```
public static function infolist(Infolist $infolist): Infolist
    {
        return $infolist
            ->schema([
                TextEntry::make('name')
                    ->columnSpanFull(),
                ViewEntry::make('status')
                    ->view('custom-view')
            ]);
    }
```

**custom-view.blade.php** (I am able to pass the course using the binding)
```
<div>
    <livewire:list-course-platform-quiz-configurations :course="$getRecord()"/>
</div>
```

And then the Livewire component class:

```
class ListCoursePlatformQuizConfigurations extends Component implements HasTable, HasForms
{
    use InteractsWithTable;
    use InteractsWithForms;

    public CatalogCourse $course;

    public function render()
    {
        return view('livewire.list-course-platform-quiz-configurations');
    }

    public function table(Table $table): Table
    {
        return $table
//            ->query(CatalogCourse::query())
            ->relationship(fn (): HasMany => $this->course->platform_quiz_configurations())
            ->columns([
                TextColumn::make('id'),
            ])
            ->filters([
                // ...
            ])
            ->actions([
                // ...
            ])
            ->bulkActions([
                // ...
            ]);
    }
}

```

Finally, the blade template **list-course-platform-quiz-configurations.blade.php**

```
<div>
    {{ $this->table }}
</div>


```
Using the above, I get the error
`App\Livewire\ListCoursePlatformQuizConfigurations::getTableRecordKey(): Return value must be of type string, null returned` which happens on the blade template **list-course-platform-quiz-configurations.blade.php**

I have confirmed that the relationship is working as expected by getting the related records at a breakpoint and inspecting them:
<img width="818" alt="image" src="https://github.com/filamentphp/filament/assets/1805757/b361a246-d578-4545-9c05-53a5bfa5ba3c">

That relationship is defined on the model **CatalogCourse** with

```
public function platform_quiz_configurations() : HasMany
    {
        return $this->setConnection('removed')->hasMany(PlatformQuizConfiguration::class, 'RemoteCourseId');
    }
```

**[Previous Error]**
With that, I get this error when viewing the resource:
**Property [$table] not found on component: [app.filament.resources.catalog-course-resource.pages.view-catalog-course]**

If I comment out `{{ $this->table }}` I do see the correct record being shown in the Livewire component, but that error message suggests (to me) that it's looking for the table component on my Resource, instead of within the Component.

I'm trying to follow https://filamentphp.com/docs/3.x/tables/adding-a-table-to-a-livewire-component#building-a-table-for-an-eloquent-relationship but can't see what I'm doing wrong here.  Thanks for any insight!

**Comment by leandrocfe** [comment, marked as answer]
```php
->view('custom-view')
```
```html
<!-- resources/views/custom-view.blade.php -->
<div>
    <livewire:list-course-platform-quiz-configurations />
</div>
```

**Reply by clrockwell** [reply]
Thanks @leandrocfe .  I'm not sure _where_ your first snippet should go.  I'm **very** new to Filament and there seem to be a few different places where `->view()` is used.  I'll try this out but if you could be more specific I would greatly appreciate it.  Thanks!

**Reply by clrockwell** [reply]
Sorry - I think I see now that it's supposed to be on `ViewEntry`

**Reply by clrockwell** [reply]
I now have the livewire component being rendered, but the table doesn't work.  I get the error:

Typed property App\Livewire\ListCoursePlatformQuizConfigurations::$course must not be accessed before initialization.  I've checked a few times and it looks like I'm following this example: https://filamentphp.com/docs/3.x/tables/adding-a-table-to-a-livewire-component#building-a-table-for-an-eloquent-relationship.  (Admittedly the docs for Filament confuse me a little as they seem to often leave out things.  In this case, I _think_ they are leaving out the `render` function for brevity but I'm not certain.)

**Reply by dev-federico** [reply]
Hi @clrockwell I'm facing the same "Typed property...must not be accessed before initialization." problem. Have you found any solution?

**Reply by clrockwell** [reply]
@spaceproj Yes - I updated the code in my question to reflect the full solution.  It might be best if you post your code in a new question and then link it here, I'm happy to try and help.

**Reply by dev-federico** [reply]
@clrockwell Sorry I missed your update on the question, I solved the problem in the same way. Thank you!

**Reply by iphai** [reply]
hello, I also have almost the same question, is it possible to use a query thrown using the $state variable? I tried with the script, the table doesn't appear properly. Here the code :

```
\Filament\Infolists\Components\View::make('jamkerja')
                            ->view('components.jamkerja')
                            ->getStateUsing(
                                function ($record)
                                {
                                    return JamKerja::where('perusahaan_id','=', $record->perusahaan_id)->get();
                                }
                            )
                            ->viewData(
                                    [
                                    'jam_kerja' => fn($state) => $state
                                    ]
                                )
                            ->columnSpan('full'),
```

if used this code ->viewData(['jam_kerja' => JamKerja::all() ]) , it works fine
